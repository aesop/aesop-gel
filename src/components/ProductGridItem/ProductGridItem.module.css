@import 'styles/index';

.base {
  background-color: papayawhip;
}

/* .productName { } */

.definitionList {
  display: grid;
  grid-template-columns: auto minmax(0, 1fr);

  dt {
    padding-right: 20px;
    border-bottom: 1px solid var(--color-grey-65);
    margin-bottom: 0;
  }

  dt,
  dd {
    padding-top: 14px;
    padding-bottom: 14px;
  }

  dd:nth-of-type(1) {
    padding-top: 0;
  }
}
