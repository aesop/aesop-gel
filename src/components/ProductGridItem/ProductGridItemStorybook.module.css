@import 'styles/index';

.holder {
  @media (--viewport-md) {
    width: 50%;
  }
  @media (--viewport-lg) {
    width: 460px;
  }
}
