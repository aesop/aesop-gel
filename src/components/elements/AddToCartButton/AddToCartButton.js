import React from 'react';
import cx from 'classnames';
import Button from '~/components/elements/Button';
import Loading from '~/components/elements/Loading';
import PROP_TYPES from './AddToCartButton.prop-types';
import styles from './AddToCartButton.module.css';

const AddToCartButton = ({
  className,
  dataTestRef,
  handleOnClick,
  isDisabled = false,
  isLoading = false,
  name,
  price,
  sku,
  loading,
}) => {
  const classSet = cx(styles.base, className);

  // TODO - need contentful translations for these
  const cartActionCopy = `Add to your cart â€” ${price}`;
  const notificationCopy = `${name} added to your cart`;
  const successCopy = `Added to your cart`;

  // @TODO Mock const
  const createCartEntry = () => {};
  const data = {};
  const error = {};

  const updateSuccessful = !loading && data && data.createCartEntry;
  const labelClassName = cx(styles.label, {
    [styles.showSuccessMessage]: updateSuccessful,
  });

  return (
    <Button
      alternate={true}
      className={classSet}
      dataTestRef={dataTestRef}
      disabled={loading || isDisabled}
      onClick={handleOnClick}
      title="Add to cart button"
    >
      {isLoading ? (
        <Loading
          className={styles.loading}
          data-test-ref={`${dataTestRef}_UPDATING`}
          isLight={true}
          isLoading={true}
        />
      ) : (
        <span
          className={labelClassName}
          data-test-ref={
            updateSuccessful ? `${dataTestRef}_SUCCESS` : undefined
          }
        >
          <span>{successCopy}</span>
          <span>{cartActionCopy}</span>
        </span>
      )}
    </Button>
  );
};

AddToCartButton.propTypes = PROP_TYPES;

export default AddToCartButton;
