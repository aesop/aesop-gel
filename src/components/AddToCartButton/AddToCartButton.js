import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useAddToCartContext, useVariantSelectContext } from '~/contexts';
import { HYPERLINK_STYLE_TYPES } from '~/constants';
import { Button } from '~/components/Button';
import { Loading } from '~/components/Loading';
import { Hyperlink } from '~/components/Hyperlink';
import styles from './AddToCartButton.module.css';

const AddToCartButton = ({
  className,
  copy,
  dataTestRef,
  isEnabled,
  isFullWidth,
  theme,
}) => {
  const addToCartContext = useAddToCartContext();
  const { selectedVariant } = useVariantSelectContext();

  if (!selectedVariant) return null;

  const {
    alternateAction,
    isInStock,
    isSellable,
    price,
    sku,
  } = selectedVariant;

  const classSet = cx(
    styles.base,
    { [styles.fullWidth]: isFullWidth },
    className,
  );

  const handleOnClick = () => {
    const { actionTypes, dispatch, onClick } = addToCartContext;

    onClick(sku, dispatch, actionTypes);
  };

  if (!!alternateAction?.url && !!alternateAction?.label) {
    return (
      <Hyperlink
        className={classSet}
        dataTestRef={dataTestRef}
        isAlternate={true}
        style={HYPERLINK_STYLE_TYPES.INTERNAL_NO_ICON_BUTTON_LINK}
        theme={theme}
        title={alternateAction?.label}
        url={alternateAction?.url}
      >
        {alternateAction?.label}
      </Hyperlink>
    );
  }

  const {
    errorMessage,
    hasError,
    isLoading,
    isUpdateSuccessful,
  } = addToCartContext;

  if (hasError) {
    /** @TODO Handle errors thrown by handleOnClick */
    console.error('Add To Cart button updateError: ', errorMessage); // eslint-disable-line
  }

  const cartActionLabel = `${copy.cartAction} â€” ${price}`;
  const updateNotificationLabel = copy.updateNotification;
  const shouldShowUpdateSuccessMessage = !isLoading && isUpdateSuccessful;

  const isButtonEnabled =
    !isLoading &&
    !!price &&
    !!sku &&
    isInStock &&
    !!sku &&
    isEnabled &&
    !hasError &&
    isSellable;

  const labelClassName = cx(
    styles.label,
    { [styles.hideLabel]: isLoading },
    { [styles.showSuccessMessage]: shouldShowUpdateSuccessMessage },
  );

  return (
    <Button
      className={classSet}
      dataTestRef={dataTestRef}
      isAlternate={true}
      isEnabled={isButtonEnabled}
      onClick={handleOnClick}
      theme={theme}
      title={isInStock ? cartActionLabel : copy.outOfStock?.title}
    >
      {isLoading && (
        <Loading
          className={styles.loading}
          data-test-ref={`${dataTestRef}_UPDATING`}
          isLoading={true}
          theme={theme === 'dark' ? 'light' : 'dark'}
        />
      )}
      <span
        className={labelClassName}
        data-test-ref={
          shouldShowUpdateSuccessMessage ? `${dataTestRef}_SUCCESS` : undefined
        }
      >
        <span>{updateNotificationLabel}</span>
        <span>{isInStock ? cartActionLabel : copy.outOfStock?.label}</span>
      </span>
    </Button>
  );
};

AddToCartButton.propTypes = {
  className: PropTypes.string,
  copy: PropTypes.shape({
    cartAction: PropTypes.string,
    updateNotification: PropTypes.string,
    outOfStock: PropTypes.shape({
      label: PropTypes.string,
      title: PropTypes.string,
    }),
  }),
  dataTestRef: PropTypes.string.isRequired,
  isEnabled: PropTypes.bool,
  isFullWidth: PropTypes.bool,
  theme: PropTypes.oneOf(['dark', 'light']),
};

AddToCartButton.defaultProps = {
  className: undefined,
  copy: {
    cartAction: undefined,
    updateNotification: undefined,
    outOfStock: {
      label: undefined,
      title: undefined,
    },
  },
  dataTestRef: 'ADD_TO_CART',
  isEnabled: true,
  isFullWidth: true,
  theme: 'dark',
};

export { AddToCartButton };
