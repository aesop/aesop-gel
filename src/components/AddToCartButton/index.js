/* istanbul ignore file */
import AddToCartButton from './AddToCartButton';

export default AddToCartButton;
