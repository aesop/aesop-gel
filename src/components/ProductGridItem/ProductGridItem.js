import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { useProductDetailContext, useVariantSelectContext } from '~/contexts';
import AddToCartButton from '~/components/AddToCartButton';
import DefinitionList from '~/components/DefinitionList';

import styles from './ProductGridItem.module.css';

const ProductGridItem = ({ className, copy, theme }) => {
  // const {
  //   selectedVariant,
  //   onVariantChange,
  //   variants,
  // } = useVariantSelectContext();

  const { productDetail } = useProductDetailContext();
  const { definitionList } = productDetail;

  const classSet = cx(styles.base, className);

  const ADD_TO_CART_BUTTON_DATA_TEST_REF = 'PRODUCT_GRID_ITEM_ADD_TO_CART_CTA';

  return (
    <div className={classSet}>
      <h3>Product grid item - mobile first</h3>
      <p>Image</p>
      <p>Product title</p>
      <p>radio buttons</p>

      <AddToCartButton
        className={styles.addToCartButton}
        copy={copy.addToCart}
        dataTestRef={ADD_TO_CART_BUTTON_DATA_TEST_REF}
        isFullWidth={true}
        theme={theme}
      />

      <DefinitionList
        className={styles.definitionList}
        hasBottomBorder={true}
        items={definitionList}
        theme={theme}
      />
    </div>
  );
};

ProductGridItem.propTypes = {
  className: PropTypes.string,
  copy: PropTypes.shape({
    addToCart: PropTypes.shape({
      cartAction: PropTypes.string,
      updateNotification: PropTypes.string,
      outOfStock: PropTypes.shape({
        label: PropTypes.string,
        title: PropTypes.string,
      }),
    }),
  }),
  theme: PropTypes.oneOf(['dark', 'light']),
};

ProductGridItem.defaultProps = {
  className: undefined,
  copy: {
    addToCart: {
      cartAction: undefined,
      updateNotification: undefined,
      outOfStock: {
        label: undefined,
        title: undefined,
      },
    },
  },
  theme: 'dark',
};

export default ProductGridItem;
