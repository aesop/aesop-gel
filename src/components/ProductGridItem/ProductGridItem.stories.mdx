import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import ProductGridItem from './ProductGridItem';
import ProductGridItemFixture from './ProductGridItem.fixture';

<Meta title="Components/ProductGridItem" component={ProductGridItem} />

# ProductGridItem

<Preview>
  <Story name="Base component">
    <ProductGridItem />
  </Story>
</Preview>

## API

<Props of={ProductGridItem} />
