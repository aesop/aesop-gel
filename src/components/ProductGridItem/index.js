/* istanbul ignore file */
import { ProductGridItem } from './ProductGridItem';

export { ProductGridItem };
