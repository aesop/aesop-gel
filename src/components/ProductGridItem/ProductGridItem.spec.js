import React from 'react';
import { render } from '@testing-library/react';
import {
  AddToCartContextProvider,
  ProductDetailContextProvider,
  VariantSelectContextProvider,
} from '~/contexts';
import ProductDetailHeaderFixture from '~/components/ProductDetailHeader/ProductDetailHeader.fixture';
import { AddToCartButtonFixture } from '~/components/AddToCartButton/AddToCartButton.fixture';
import { mockAddToCartButtonOnClick } from '~/components/AddToCartButton/__mocks__/AddToCartButton.onClick';
import ProductGridItem from './ProductGridItem';
import ProductGridItemFixture from './ProductGridItem.fixture';

describe('<ProductGridItem />', () => {
  it('should be defined', () => {
    expect(ProductGridItem).toBeDefined();
  });

  it('renders base component correctly', () => {
    const { container } = render(
      <AddToCartContextProvider onClick={mockAddToCartButtonOnClick}>
        <ProductDetailContextProvider
          product={ProductDetailHeaderFixture.product}
        >
          <VariantSelectContextProvider
            variants={ProductGridItemFixture.variantOptions}
          >
            <ProductGridItem
              copy={{
                addToCart: AddToCartButtonFixture.copy,
                ...ProductGridItemFixture.copy,
              }}
              info={ProductGridItemFixture.info}
              url={ProductGridItemFixture.url}
            />
          </VariantSelectContextProvider>
        </ProductDetailContextProvider>
      </AddToCartContextProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
