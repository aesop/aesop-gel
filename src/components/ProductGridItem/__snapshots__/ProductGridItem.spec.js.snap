// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductGridItem /> renders base component correctly 1`] = `
<div
  className="base dark"
>
  <a
    className="base left dark imageLink fadeEnter"
    href="/p/"
    target="_self"
  >
    <picture
      className="base picture dark image"
    >
      <img />
    </picture>
  </a>
  <h5
    className="base mediumWeightFont xXSmall dark productName"
  >
    <a
      className="base left dark productNameLink"
      href="/p/"
      target="_self"
    >
      Lorem ipsum dolor
    </a>
  </h5>
  <div
    className="infoVariantHolder"
  >
    <div
      className="infoHolder"
    >
      <p
        className="base dark info"
      >
        4 Sizes / From $ 26.45
      </p>
    </div>
    <ul
      className="base variants"
    >
      <li
        className="radio"
      >
        <label
          className="label"
          htmlFor="option-ARD33"
        >
          <input
            aria-checked={true}
            checked={true}
            className="input"
            data-test-ref="PRODUCT_GRID_ITEM_VARIANT_SELECT"
            id="option-ARD33"
            name="sku"
            onChange={[Function]}
            tabIndex={0}
            type="radio"
            value="ARD33"
          />
          <span
            className="pot dark"
          />
          <span
            className="labelContent dark"
          >
            50 mL - In stock
          </span>
        </label>
      </li>
      <li
        className="radio"
      >
        <label
          className="label"
          htmlFor="option-ARD32"
        >
          <input
            aria-checked={false}
            checked={false}
            className="input"
            data-test-ref="PRODUCT_GRID_ITEM_VARIANT_SELECT"
            id="option-ARD32"
            name="sku"
            onChange={[Function]}
            tabIndex={0}
            type="radio"
            value="ARD32"
          />
          <span
            className="pot dark"
          />
          <span
            className="labelContent dark"
          >
            100 mL - Not in stock
          </span>
        </label>
      </li>
      <li
        className="radio"
      >
        <label
          className="label"
          htmlFor="option-ARD30"
        >
          <input
            aria-checked={false}
            checked={false}
            className="input"
            data-test-ref="PRODUCT_GRID_ITEM_VARIANT_SELECT"
            id="option-ARD30"
            name="sku"
            onChange={[Function]}
            tabIndex={0}
            type="radio"
            value="ARD30"
          />
          <span
            className="pot dark"
          />
          <span
            className="labelContent dark"
          >
            180 mL - In stock
          </span>
        </label>
      </li>
      <div
        className="errorMessage"
      >
        
      </div>
    </ul>
  </div>
  <button
    className="base alternate blockStyle dark base fullWidth addToCartButton"
    data-test-ref="ADD_TO_CART"
    disabled={false}
    onClick={[Function]}
    title="Add to your cart — $26.45"
    type="button"
  >
    <span
      className="label"
    >
      <span>
        Added to your cart
      </span>
      <span>
        Add to your cart — $26.45
      </span>
    </span>
  </button>
  <dl
    className="base dark hasBottomBorder definitionList"
  >
    <dt
      className="term slideIn"
    >
      Lorem ipsum dolor
    </dt>
    <dd
      className="description slideIn"
    >
      Lorem ipsum dolort, consectetur adipiscing elit.
    </dd>
    <dt
      className="term slideIn"
    >
      Ut consectetur mi
    </dt>
    <dd
      className="description slideIn"
    >
      Ut consectetur, vitae libero imperdiet id. 
    </dd>
  </dl>
</div>
`;
